src/tools/gravity-parseList.c

```diff
@@ -45,6 +39,96 @@
 // Number of invalid domains to print before skipping the rest
 #define MAX_INVALID_DOMAINS 5
 
+// Validate domain name
+static inline bool __attribute__((pure)) valid_domain(const char *domain, const size_t len)
+{
+	// Domain must not be NULL or empty
+	if(domain == NULL || len == 0)
+		return false;
+
+	// Domain must not start or end with a hyphen or dot
```

Comment:
Not sure about that. Isn't a final dot just indicating the root zone?

Reply:
Yes, but this will never be (and never has been) matched. `dnsmasq` drops any possibly existing final dots.

Reply:
So if `dnsmasq` receives an query to `somedomain.com.` it will convert it to `somedomain.com` and send this upstream? In that case, we should not reject adlist entries with `somedomain.com.` but instead convert it to `somedomain.com` and add it to gravity. Otherwiese `somedomain.com.` would bypass Pi-hole even if `somedomain.com.` is blocked.

(And this is actually happening. Added `google.com.` to blacklist

```
Oct 15 20:46:05 dnsmasq[111250]: query[A] google.com from 10.0.1.59
Oct 15 20:46:05 dnsmasq[111250]: cached google.com is 172.217.18.110

```
)